File: ios/TougeDash/Telemetry/ECUControlCoordinator.swift
import Combine
import Foundation
import UIKit

enum ECUControlKind: Equatable, Sendable {
    case switchValue
    case rotary
}

struct PendingECUControl: Equatable, Sendable {
    let kind: ECUControlKind
    let channel: Int
    let targetSnapshot: ECUControlSnapshot
    let startedAt: Date
    let loopbackRevisionAtStart: UInt64
}

@MainActor
final class ECUControlCoordinator: ObservableObject {
    typealias Writer = (Data, @escaping (Result<Void, Error>) -> Void) -> Void

    @Published private(set) var observedSnapshot: ECUControlSnapshot?
    @Published private(set) var pending: PendingECUControl?
    @Published private(set) var isConnected = false
    @Published private(set) var transportAvailable = false
    @Published private(set) var hasSynchronizedLoopback = false
    @Published private(set) var errorMessage: String?

    var writer: Writer?

    private var loopback = ECUControlLoopbackAccumulator()
    private var confirmationTask: Task<Void, Never>?
    private var isApplicationActive: Bool
    private let notificationCenter: NotificationCenter
    nonisolated(unsafe) private var observers: [NSObjectProtocol] = []

    init(notificationCenter: NotificationCenter = .default, applicationIsActive: Bool? = nil) {
        self.notificationCenter = notificationCenter
        isApplicationActive = applicationIsActive ?? (UIApplication.shared.applicationState == .active)
        observers.append(notificationCenter.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in self?.applicationDidBecomeInactive() }
        })
        observers.append(notificationCenter.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            Task { @MainActor in self?.applicationDidBecomeActive() }
        })
    }

    deinit {
        confirmationTask?.cancel()
        for observer in observers { notificationCenter.removeObserver(observer) }
    }

    var isReady: Bool {
        isConnected && transportAvailable && hasSynchronizedLoopback && isApplicationActive && pending == nil
    }

    func connectionChanged(isConnected: Bool) {
        self.isConnected = isConnected
        transportAvailable = false
        observedSnapshot = nil
        hasSynchronizedLoopback = false
        pending = nil
        errorMessage = nil
        loopback.reset()
        confirmationTask?.cancel()
        confirmationTask = nil
    }

    func transportAvailabilityChanged(_ available: Bool) {
        transportAvailable = isConnected && available
        if !transportAvailable {
            failPending(localized("Kanał sterowania ECU jest niedostępny."))
        }
    }

    func ingest(_ frame: EMUFrame, receivedAt: Date = .now) {
        guard isConnected, loopback.apply(frame, receivedAt: receivedAt) else { return }
        refreshLoopbackState()
    }

    func ingestStatusFrame(_ data: Data) {
        guard isConnected, loopback.applyStatusFrame(data) else { return }
        refreshLoopbackState()
    }

    func switchValue(channel: Int) -> Bool? {
        guard observedSnapshot != nil || pending != nil else { return nil }
        return displaySnapshot.switchValue(channel: channel)
    }

    func rotaryValue(channel: Int) -> Int? {
        guard observedSnapshot != nil || pending != nil else { return nil }
        return displaySnapshot.rotaryValue(channel: channel).map(Int.init)
    }

    func isPending(kind: ECUControlKind, channel: Int) -> Bool {
        pending?.kind == kind && pending?.channel == channel
    }

    @discardableResult
    func toggleSwitch(channel: Int) -> Bool {
        guard isReady, let observedSnapshot,
              let current = observedSnapshot.switchValue(channel: channel),
              let target = observedSnapshot.settingSwitch(channel: channel, to: !current) else {
            return false
        }
        return send(kind: .switchValue, channel: channel, target: target)
    }

    @discardableResult
    func setRotary(channel: Int, value: Int) -> Bool {
        guard isReady, let observedSnapshot,
              observedSnapshot.rotaryValue(channel: channel) != UInt8(value),
              let target = observedSnapshot.settingRotary(channel: channel, to: value) else {
            return false
        }
        return send(kind: .rotary, channel: channel, target: target)
    }

    var availabilityLabel: String {
        if !isConnected { return localized("BRAK POŁĄCZENIA") }
        if !transportAvailable { return localized("TYLKO ODCZYT") }
        if !isApplicationActive { return localized("APLIKACJA W TLE") }
        if !hasSynchronizedLoopback {
            let channels = loopback.missingChannels.map(String.init).joined(separator: ", ")
            return String(format: localized("CZEKAM NA KANAŁY %@"), channels)
        }
        if pending != nil { return localized("OCZEKIWANIE NA EMU") }
        if errorMessage != nil { return localized("BŁĄD STEROWANIA") }
        return localized("POTWIERDZONE PRZEZ EMU")
    }

    private var displaySnapshot: ECUControlSnapshot {
        pending?.targetSnapshot ?? observedSnapshot ?? ECUControlSnapshot()
    }

    private func send(kind: ECUControlKind, channel: Int, target: ECUControlSnapshot) -> Bool {
        guard let writer else { return false }
        errorMessage = nil
        let request = PendingECUControl(
            kind: kind,
            channel: channel,
            targetSnapshot: target,
            startedAt: .now,
            loopbackRevisionAtStart: loopback.currentRevision
        )
        pending = request
        let frame = target.encodedStatusFrame()

        writer(frame) { [weak self] result in
            Task { @MainActor in
                guard let self, self.pending == request else { return }
                if case .failure(let error) = result {
                    self.failPending(String(format: localized("Nie udało się wysłać zmiany: %@"), error.localizedDescription))
                }
            }
        }

        confirmationTask?.cancel()
        confirmationTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(2))
            guard !Task.isCancelled, let self, self.pending == request else { return }
            self.failPending(localized("EMU nie potwierdziło zmiany w wymaganym czasie."))
        }
        return true
    }

    private func refreshLoopbackState() {
        guard let snapshot = loopback.synchronizedSnapshot() else {
            hasSynchronizedLoopback = false
            return
        }
        observedSnapshot = snapshot
        hasSynchronizedLoopback = true

        let confirmedSnapshot = pending.flatMap {
            loopback.snapshotConfirming(
                kind: $0.kind,
                channel: $0.channel,
                afterRevision: $0.loopbackRevisionAtStart
            )
        }
        if let pending, let confirmedSnapshot, confirms(pending, with: confirmedSnapshot) {
            self.pending = nil
            errorMessage = nil
            confirmationTask?.cancel()
            confirmationTask = nil
        }
    }

    private func confirms(_ pending: PendingECUControl, with snapshot: ECUControlSnapshot) -> Bool {
        switch pending.kind {
        case .switchValue:
            return snapshot.switchValue(channel: pending.channel) == pending.targetSnapshot.switchValue(channel: pending.channel)
        case .rotary:
            return snapshot.rotaryValue(channel: pending.channel) == pending.targetSnapshot.rotaryValue(channel: pending.channel)
        }
    }

    private func failPending(_ message: String) {
        guard pending != nil else { return }
        pending = nil
        confirmationTask?.cancel()
        confirmationTask = nil
        errorMessage = message
    }

    private func applicationDidBecomeInactive() {
        isApplicationActive = false
        failPending(localized("Sterowanie przerwano po przejściu aplikacji w tło."))
    }

    private func applicationDidBecomeActive() {
        isApplicationActive = true
        refreshLoopbackState()
    }
}

File: android/app/src/main/java/it/letscode/tougedash/telemetry/EcuControlCoordinator.kt
package it.letscode.tougedash.telemetry

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class EcuControlKind { SWITCH, ROTARY }
enum class EcuControlError { WRITE_REJECTED, CONFIRMATION_TIMEOUT, BACKGROUND, TRANSPORT_LOST }

data class PendingEcuControl(
    val kind: EcuControlKind,
    val channel: Int,
    val target: EcuControlSnapshot,
    val startedAt: Long,
    val loopbackRevisionAtStart: Long
)

data class EcuControlState(
    val observed: EcuControlSnapshot? = null,
    val displayed: EcuControlSnapshot? = null,
    val pending: PendingEcuControl? = null,
    val connected: Boolean = false,
    val transportAvailable: Boolean = false,
    val synchronizedLoopback: Boolean = false,
    val missingLoopbackChannels: List<Int> = listOf(252, 253, 254),
    val applicationActive: Boolean = true,
    val error: EcuControlError? = null
) {
    val ready: Boolean get() = connected && transportAvailable && synchronizedLoopback && applicationActive && pending == null
}

class EcuControlCoordinator(private val scope: CoroutineScope) {
    private val loopback = EcuControlLoopbackAccumulator()
    private val mutableState = MutableStateFlow(EcuControlState())
    val state = mutableState.asStateFlow()
    private var writer: ((ByteArray) -> Boolean)? = null
    private var confirmationJob: Job? = null

    @Synchronized
    fun connectionChanged(connected: Boolean) {
        writer = null
        confirmationJob?.cancel()
        confirmationJob = null
        loopback.reset()
        mutableState.value = EcuControlState(connected = connected, applicationActive = mutableState.value.applicationActive)
    }

    @Synchronized
    fun transportChanged(available: Boolean, writer: ((ByteArray) -> Boolean)? = null) {
        this.writer = writer.takeIf { available }
        val current = mutableState.value
        mutableState.value = current.copy(
            transportAvailable = current.connected && available,
            pending = if (available) current.pending else null,
            displayed = if (available) current.displayed else current.observed,
            error = if (!available && current.pending != null) EcuControlError.TRANSPORT_LOST else current.error
        )
    }

    @Synchronized
    fun applicationActive(active: Boolean) {
        val current = mutableState.value
        if (!active && current.pending != null) confirmationJob?.cancel()
        mutableState.value = current.copy(
            applicationActive = active,
            pending = if (active) current.pending else null,
            displayed = if (active) current.displayed else current.observed,
            error = if (!active && current.pending != null) EcuControlError.BACKGROUND else current.error
        )
        if (active) refresh()
    }

    @Synchronized
    fun ingest(frame: EmuFrame, receivedAt: Long = System.currentTimeMillis()) {
        if (!mutableState.value.connected || !loopback.apply(frame, receivedAt)) return
        refresh()
    }

    @Synchronized
    fun toggleSwitch(channel: Int): Boolean {
        val current = mutableState.value
        val observed = current.observed ?: return false
        if (!current.ready) return false
        val value = observed.switchValue(channel) ?: return false
        val target = observed.settingSwitch(channel, !value) ?: return false
        return send(EcuControlKind.SWITCH, channel, target)
    }

    @Synchronized
    fun setRotary(channel: Int, value: Int): Boolean {
        val current = mutableState.value
        val observed = current.observed ?: return false
        if (!current.ready || observed.rotaryValue(channel) == value) return false
        val target = observed.settingRotary(channel, value) ?: return false
        return send(EcuControlKind.ROTARY, channel, target)
    }

    @Synchronized
    fun transportWriteFailed() {
        val current = mutableState.value
        if (current.pending == null) return
        confirmationJob?.cancel()
        confirmationJob = null
        mutableState.value = current.copy(pending = null, displayed = current.observed, error = EcuControlError.WRITE_REJECTED)
    }

    private fun send(kind: EcuControlKind, channel: Int, target: EcuControlSnapshot): Boolean {
        val currentWriter = writer ?: return false
        val request = PendingEcuControl(kind, channel, target, System.currentTimeMillis(), loopback.currentRevision)
        mutableState.value = mutableState.value.copy(pending = request, displayed = target, error = null)
        if (!currentWriter(target.encodeStatusFrame())) {
            mutableState.value = mutableState.value.copy(pending = null, displayed = mutableState.value.observed, error = EcuControlError.WRITE_REJECTED)
            return false
        }
        confirmationJob?.cancel()
        confirmationJob = scope.launch {
            delay(2_000)
            synchronized(this@EcuControlCoordinator) {
                val current = mutableState.value
                if (current.pending == request) {
                    mutableState.value = current.copy(pending = null, displayed = current.observed, error = EcuControlError.CONFIRMATION_TIMEOUT)
                }
            }
        }
        return true
    }

    private fun refresh() {
        val snapshot = loopback.synchronizedSnapshot()
        val current = mutableState.value
        if (snapshot == null) {
            mutableState.value = current.copy(
                synchronizedLoopback = false,
                missingLoopbackChannels = loopback.missingChannels
            )
            return
        }
        val confirmed = current.pending?.takeIf {
            val confirmation = loopback.snapshotConfirming(it.kind, it.channel, it.loopbackRevisionAtStart)
            when (it.kind) {
                EcuControlKind.SWITCH -> confirmation?.switchValue(it.channel) == it.target.switchValue(it.channel)
                EcuControlKind.ROTARY -> confirmation?.rotaryValue(it.channel) == it.target.rotaryValue(it.channel)
            }
        }
        if (confirmed != null) {
            confirmationJob?.cancel()
            confirmationJob = null
        }
        mutableState.value = current.copy(
            observed = snapshot,
            displayed = if (confirmed != null || current.pending == null) snapshot else current.displayed,
            pending = if (confirmed != null) null else current.pending,
            synchronizedLoopback = true,
            missingLoopbackChannels = emptyList(),
            error = if (confirmed != null) null else current.error
        )
    }
}
